#Problem Statement

This dataset is created for the prediction of House Prices from an Indian perspective Towns. The dataset contains various features that are important during the application for the selling the Houses. The predicted output obtained from the classification algorithm gives a fair idea about the chances of a Houses Sold.


<table align="center" width=100%>
    <tr>
        </td>
            </td>
        </td>
        <td>
            <div align="center">
                <font color="#216180" size=8px>
                    <b> Regression modeling on House Price Data
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [1]:
# import Pandas
import pandas as pd

#import numpy
import numpy as np

#import subpackage of Matpiotlib
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap

#import Seaborn
import seaborn as sns

# # to supress warnings
# from warnings import filtervarnings. filtersarnings('ignore')



### Load the data state

In [2]:
# Load the data and name it 'df_House_price'

In [3]:
# Load the csv file
# store the data in 'df_houseprice'
df_houseprice =pd.read_csv(r"C:\Users\HP\Downloads\House_Price1.csv")

 #display first file observation                         
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [4]:
df=pd.read_csv(r"C:\Users\HP\Downloads\House_Price1.csv")
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

## Data Overview

In [5]:
# Name the columns 
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

**Interpretation**

 - `Price`:Price of the house.
 -  `resid_area`:Residential area where my house is located.
   

In [6]:
# Basic Info of the Dataset
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


**Interpretation**

- We have 13 numerical columns and 3 categorical columns.
- We have null values in the data.
- We have total 16 columns.

In [7]:
# Shape of the data
df_houseprice.shape

(506, 16)

**Interpretation**
- we have total 16 columns
- we have 506 observations

In [8]:
df_houseprice.describe().T

count       mean        std        min        25%        50%  \
price        506.0  22.528854   9.182176   5.000000  17.025000  21.200000   
resid_area   506.0  41.136779   6.860353  30.460000  35.190000  39.690000   
air_qual     506.0   0.554695   0.115878   0.385000   0.449000   0.538000   
room_num     506.0   6.284634   0.702617   3.561000   5.885500   6.208500   
age          506.0  68.574901  28.148861   2.900000  45.025000  77.500000   
distance     506.0   3.971996   2.108532   1.130000   2.270000   3.385000   
teachers     506.0  21.544466   2.164946  18.000000  19.800000  20.950000   
poor_prop    506.0  12.653063   7.141062   1.730000   6.950000  11.360000   
n_hos_beds   498.0   7.899767   1.476683   5.268000   6.634500   7.999000   
n_hot_rooms  506.0  13.041605   5.238957  10.057600  11.189800  12.720000   
rainfall     506.0  39.181818  12.513697   3.000000  28.000000  39.000000   
parks        506.0   0.054454   0.010632   0.033292   0.046464   0.053507   
Sold         506.0   0.454545   0.498422   0.000000   0.000000   0.000000   

                   75%         max  
price        25.000000   50.000000  
resid_area   48.100000   57.740000  
air_qual      0.624000    0.871000  
room_num      6.623500    8.780000  
age          94.075000  100.000000  
distance      5.367500   12.320000  
teachers     22.600000   27.400000  
poor_prop    16.955000   37.970000  
n_hos_beds    9.088000   10.876000  
n_hot_rooms  14.170800  101.120000  
rainfall     50.000000   60.000000  
parks         0.061397    0.086711  
Sold          1.000000    1.000000

**Interpretation**
- We have outliers in the data, most of the outliers are in age column.

## Data prerocessing.


### a.datacleaning

In [9]:
# show the first five observations.
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [10]:
# Rename the columns
# Step 1:Show the column names.
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [11]:
# Step 2:Allocate new column names to old column names.
# Make the dictionary {'Old column name':'New column name'}
d1={'price':'Price',
  'resid_area':'Resid_Area',
  'air_qual':'Air_Qual',
   'room_num':'Room_Nums',
    'age':'Age',
    'distance':'Distance',
    'teachers':'Teachers',
    'poor_prop':'Preportion_of_poor_popuation',
    'airport':'Airport',
    'n_hos_beds':'Number_of_hospital_beds',
    'n_hot_rooms':'Number_of_hetel_rooms',
     'waterbody': 'Waterbody',
    'rainfall':'Rainfall',
    'bus_ter':'Bus_Terminal',
    'parks':'Parks',
    'Sold':'sold'
 }

In [12]:
df_houseprice=df_houseprice.rename(columns=d1)


In [13]:
# Authenticate the data
df_houseprice.columns

Index(['Price', 'Resid_Area', 'Air_Qual', 'Room_Nums', 'Age', 'Distance',
       'Teachers', 'Preportion_of_poor_popuation', 'Airport',
       'Number_of_hospital_beds', 'Number_of_hetel_rooms', 'Waterbody',
       'Rainfall', 'Bus_Terminal', 'Parks', 'sold'],
      dtype='object')

### b.Null value handling

In [14]:
# Find the valid null values
df_houseprice.isnull().sum()

Price                             0
Resid_Area                        0
Air_Qual                          0
Room_Nums                         0
Age                               0
Distance                          0
Teachers                          0
Preportion_of_poor_popuation      0
Airport                           0
Number_of_hospital_beds           8
Number_of_hetel_rooms             0
Waterbody                       155
Rainfall                          0
Bus_Terminal                      0
Parks                             0
sold                              0
dtype: int64

In [15]:
# percentage missing value
round(df_houseprice.isnull().sum()/len(df_houseprice)*100,3)

Price                            0.000
Resid_Area                       0.000
Air_Qual                         0.000
Room_Nums                        0.000
Age                              0.000
Distance                         0.000
Teachers                         0.000
Preportion_of_poor_popuation     0.000
Airport                          0.000
Number_of_hospital_beds          1.581
Number_of_hetel_rooms            0.000
Waterbody                       30.632
Rainfall                         0.000
Bus_Terminal                     0.000
Parks                            0.000
sold                             0.000
dtype: float64

In [16]:
# Find the nature of the column
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Price                         506 non-null    float64
 1   Resid_Area                    506 non-null    float64
 2   Air_Qual                      506 non-null    float64
 3   Room_Nums                     506 non-null    float64
 4   Age                           506 non-null    float64
 5   Distance                      506 non-null    float64
 6   Teachers                      506 non-null    float64
 7   Preportion_of_poor_popuation  506 non-null    float64
 8   Airport                       506 non-null    object 
 9   Number_of_hospital_beds       498 non-null    float64
 10  Number_of_hetel_rooms         506 non-null    float64
 11  Waterbody                     351 non-null    object 
 12  Rainfall                      506 non-null    int64  
 13  Bus_T

In [17]:
# This is a categorical column so we can replace the null with mode.
df_houseprice.Waterbody.mode()

0    River
Name: Waterbody, dtype: object

In [18]:
# Replace the null in that column
df_houseprice.Waterbody=df_houseprice.Waterbody.fillna('River')

In [19]:
# In Number_of_hospital_beds contains less than 10% missing records
df_houseprice.dropna(inplace=True)

In [20]:
round(df_houseprice.isnull().sum()/len(df_houseprice)*100,3)

Price                           0.0
Resid_Area                      0.0
Air_Qual                        0.0
Room_Nums                       0.0
Age                             0.0
Distance                        0.0
Teachers                        0.0
Preportion_of_poor_popuation    0.0
Airport                         0.0
Number_of_hospital_beds         0.0
Number_of_hetel_rooms           0.0
Waterbody                       0.0
Rainfall                        0.0
Bus_Terminal                    0.0
Parks                           0.0
sold                            0.0
dtype: float64

In [21]:
## Prepare Data for Machine Learning
# We know that machine learning algorithms are mathematical expressions, so we need to provide the data to the algorithm in specific format.
# Encoding 
# Step 1 divide the data based on their datatypes
df_cat=df_houseprice.select_dtypes(include =['object'])
df_cat.columns


Index(['Airport', 'Waterbody', 'Bus_Terminal'], dtype='object')

In [22]:
# Use Dummy Encoding
dummy_var = pd.get_dummies(data =df_cat, drop_first = True)
dummy_var=dummy_var.reset_index()

In [23]:
# Use Dummy Encoding
dummy_var2 = pd.get_dummies(data = df_cat)
dummy_var2

Airport_NO  Airport_YES  Waterbody_Lake  Waterbody_Lake and River  \
0         False         True           False                     False   
1          True        False            True                     False   
2          True        False           False                     False   
3         False         True            True                     False   
4          True        False            True                     False   
..          ...          ...             ...                       ...   
501        True        False           False                      True   
502       False         True           False                      True   
503        True        False           False                     False   
504       False         True           False                     False   
505       False         True           False                     False   

     Waterbody_River  Bus_Terminal_YES  
0               True              True  
1              False              True  
2               True              True  
3              False              True  
4              False              True  
..               ...               ...  
501            False              True  
502            False              True  
503             True              True  
504             True              True  
505             True              True  

[498 rows x 6 columns]

In [24]:
df_num=df_houseprice.select_dtypes(include =[np.number])
df_num.head()

Price  Resid_Area  Air_Qual  Room_Nums   Age  Distance  Teachers  \
0   24.0       32.31     0.538      6.575  65.2      4.35      24.7   
1   21.6       37.07     0.469      6.421  78.9      4.99      22.2   
2   34.7       37.07     0.469      7.185  61.1      5.03      22.2   
3   33.4       32.18     0.458      6.998  45.8      6.21      21.3   
4   36.2       32.18     0.458      7.147  54.2      6.16      21.3   

   Preportion_of_poor_popuation  Number_of_hospital_beds  \
0                          4.98                    5.480   
1                          9.14                    7.332   
2                          4.03                    7.394   
3                          2.94                    9.268   
4                          5.33                    8.824   

   Number_of_hetel_rooms  Rainfall     Parks  sold  
0                11.1920        23  0.049347     0  
1                12.1728        42  0.046146     1  
2               101.1200        38  0.045764     0  
3                11.2672        45  0.047151     0  
4                11.2896        55  0.039474     0

In [25]:
from sklearn.preprocessing import StandardScaler

In [26]:
df_target=df_num['Price']
df_num=df_num.drop('Price', axis = 1)


In [27]:
# Initiate the standard scaler
ss=StandardScaler()

In [28]:
# Let us transform the data
Scale=ss.fit_transform(df_num)

In [29]:
Scale

array([[-1.28244435, -0.13709359,  0.40897478, ..., -1.30206786,
        -0.47313205, -0.91152748],
       [-0.59005882, -0.73319492,  0.18997385, ...,  0.21519397,
        -0.77522463,  1.09705963],
       [-0.59005882, -0.73319492,  1.276446  , ..., -0.10422957,
        -0.81123666, -0.91152748],
       ...,
       [ 0.11687263,  0.16527665,  0.97923045, ..., -0.66322077,
         0.30293265,  1.09705963],
       [ 0.11687263,  0.16527665,  0.72041117, ...,  0.6144734 ,
         0.59750894,  1.09705963],
       [ 0.11687263,  0.16527665, -0.36606097, ...,  0.45476163,
         0.56369225,  1.09705963]])

In [30]:
df_Scale=pd.DataFrame(Scale,columns=df_num.columns)

In [31]:
df_Scale = df_Scale.reset_index()
df_Scale

index  Resid_Area  Air_Qual  Room_Nums       Age  Distance  Teachers  \
0        0   -1.282444 -0.137094   0.408975 -0.115644  0.172845  1.462167   
1        1   -0.590059 -0.733195   0.189974  0.369741  0.476100  0.304963   
2        2   -0.590059 -0.733195   1.276446 -0.260905  0.495053  0.304963   
3        3   -1.301354 -0.828226   1.010516 -0.802977  1.054179 -0.111631   
4        4   -1.301354 -0.828226   1.222407 -0.505369  1.030487 -0.111631   
..     ...         ...       ...        ...       ...       ...       ...   
493    493    0.116873  0.165277   0.434572  0.022531 -0.637413 -1.176258   
494    494    0.116873  0.165277  -0.238073  0.291796 -0.732180 -1.176258   
495    495    0.116873  0.165277   0.979230  0.798438 -0.779564 -1.176258   
496    496    0.116873  0.165277   0.720411  0.738208 -0.684797 -1.176258   
497    497    0.116873  0.165277  -0.366061  0.437057 -0.599506 -1.176258   

     Preportion_of_poor_popuation  Number_of_hospital_beds  \
0                       -1.066368                -1.640299   
1                       -0.484136                -0.384875   
2                       -1.199330                -0.342847   
3                       -1.351886                 0.927490   
4                       -1.017382                 0.626514   
..                            ...                      ...   
493                     -0.409958                 0.981720   
494                     -0.492534                -0.872945   
495                     -0.973995                -1.641654   
496                     -0.856429                 0.027273   
497                     -0.660486                 1.613499   

     Number_of_hetel_rooms  Rainfall     Parks      sold  
0                -0.352111 -1.302068 -0.473132 -0.911527  
1                -0.166019  0.215194 -0.775225  1.097060  
2                16.710340 -0.104230 -0.811237 -0.911527  
3                -0.337843  0.454762 -0.680402 -0.911527  
4                -0.333593  1.253320 -1.404723 -0.911527  
..                     ...       ...       ...       ...  
493              -0.164805 -0.982644  0.155116  1.097060  
494               0.022197 -1.541636  0.522886  1.097060  
495              -0.162528 -0.663221  0.302933  1.097060  
496               0.403791  0.614473  0.597509  1.097060  
497              -0.549435  0.454762  0.563692  1.097060  

[498 rows x 13 columns]

In [32]:
# Make a DataFrame for independant variables
df_independant=pd.concat([df_Scale, dummy_var] , axis =1)
df_independant.head()

index  Resid_Area  Air_Qual  Room_Nums       Age  Distance  Teachers  \
0      0   -1.282444 -0.137094   0.408975 -0.115644  0.172845  1.462167   
1      1   -0.590059 -0.733195   0.189974  0.369741  0.476100  0.304963   
2      2   -0.590059 -0.733195   1.276446 -0.260905  0.495053  0.304963   
3      3   -1.301354 -0.828226   1.010516 -0.802977  1.054179 -0.111631   
4      4   -1.301354 -0.828226   1.222407 -0.505369  1.030487 -0.111631   

   Preportion_of_poor_popuation  Number_of_hospital_beds  \
0                     -1.066368                -1.640299   
1                     -0.484136                -0.384875   
2                     -1.199330                -0.342847   
3                     -1.351886                 0.927490   
4                     -1.017382                 0.626514   

   Number_of_hetel_rooms  Rainfall     Parks      sold  index  Airport_YES  \
0              -0.352111 -1.302068 -0.473132 -0.911527      0         True   
1              -0.166019  0.215194 -0.775225  1.097060      1        False   
2              16.710340 -0.104230 -0.811237 -0.911527      2        False   
3              -0.337843  0.454762 -0.680402 -0.911527      3         True   
4              -0.333593  1.253320 -1.404723 -0.911527      4        False   

   Waterbody_Lake and River  Waterbody_River  
0                     False             True  
1                     False            False  
2                     False             True  
3                     False            False  
4                     False            False

# Prepare the data for training and 


In [33]:
import statsmodels 
import statsmodels.api as sm
df_independant = sm.add_constant(df_independant)

In [34]:
from sklearn.model_selection import train_test_split

In [35]:
X_train,X_test,y_train,y_test = train_test_split(df_independant,df_target,random_state =10, test_size = 0.2)

In [36]:
print('X_train',X_train.shape)
print('y_train',y_train.shape)
print('X_test',X_test.shape)
print('y_test',y_test.shape)

X_train (398, 18)
y_train (398,)
X_test (100, 18)
y_test (100,)


Linear Regression


In [37]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

#Create a Linear regression model
model=LinearRegression()

#Fit the model to the training data
model.fit(X_train, y_train)

#Predict on the testing data 
predictions = model.predict(X_test)

In [38]:
mse= mean_squared_error(y_test,predictions)
print("Mean Squared Error",mse)

Mean Squared Error 15.498182384260264


In [39]:
predictions

array([22.54643539, 27.39840735,  6.01371194, 24.62279556, 12.15893075,
       13.24147688, 16.67022346, 25.66565456,  9.50703345, 28.11704874,
       20.10380925, 31.97996861, 27.7494175 , 21.2841941 , 26.64068987,
        8.1693733 , 23.29099208, 20.32076214, 35.32087816, 33.05755077,
       37.53089866, 22.31822842, 17.90935745, 24.93288423,  3.61313364,
       22.42015712, 24.04955853,  1.14080403, 22.1528586 , 21.98095176,
       30.99864887, 18.47460094, 24.47115012, 15.94716737, 24.85276297,
       27.60021664, 23.10156351, 23.57407559, 36.07856382, 41.83009608,
       30.57095683, 12.59562389, 19.78670774, 31.98447659, 13.3137158 ,
       26.24526278, 31.66815717, 24.94906032, 19.30734382, 24.98848122,
       23.56415543, 25.95215834, 26.68869766, 23.9541815 , 21.12523709,
       25.24908744, 10.13721903, 15.01905518, 20.7488436 , 25.20136288,
       19.53847673, 26.30949469, 33.61883399, 23.20007772, 34.05337745,
       15.58818707, 22.34851113, 18.35483462, 33.93630801, 33.10

In [40]:
from sklearn.metrics import mean_absolute_error

In [41]:
mae= mean_absolute_error(y_test,predictions)
print("Mean Absolute error",mae)

Mean Absolute error 3.1331429132865867


In [42]:
rms=mse**0.5
print("root_mean_square",rms)


root_mean_square 3.9367730928084064


## Ridge Model

In [43]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a Ridge regression model
ridge_model = Ridge(alpha=1.0)  # alpha is the regularization strength

# Fit the model to the training data
ridge_model.fit(X_train, y_train)

# Predict on the testing data
predictions = ridge_model.predict(X_test)

# Calculate mean squared error
mse = mean_squared_error(y_test, predictions)
print("Mean Squared Error:", mse)



Mean Squared Error: 15.459329726840302


## Lasso Model

In [44]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_squared_error

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a Lasso regression model
lasso_model = Lasso(alpha=1.0)  # alpha is the regularization strength

# Fit the model to the training data
lasso_model.fit(X_train, y_train)

# Predict on the testing data
predictions = lasso_model.predict(X_test)

# Calculate mean squared error
mse = mean_squared_error(y_test, predictions)
print("Mean Squared Error:", mse)


Mean Squared Error: 20.698379031513387


## Decision Tree Regressor

In [45]:
from sklearn.tree import DecisionTreeRegressor


# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a Decision Tree regression model
tree_model = DecisionTreeRegressor()

# Fit the model to the training data
tree_model.fit(X_train, y_train)

# Predict on the testing data
predictions = tree_model.predict(X_test)

# Calculate mean squared error
mse = mean_squared_error(y_test, predictions)
print("Mean Squared Error:", mse)

Mean Squared Error: 11.673900000000001


## Random Forest Regressor

In [46]:
from sklearn.ensemble import RandomForestRegressor


# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a Random Forest regression model
forest_model = RandomForestRegressor(n_estimators=100, random_state=42)

# Fit the model to the training data
forest_model.fit(X_train, y_train)

# Predict on the testing data
predictions = forest_model.predict(X_test)

# Calculate mean squared error
mse = mean_squared_error(y_test, predictions)
print("Mean Squared Error:", mse)


Mean Squared Error: 6.433220049999999


## Gaussian Process Regressor

In [47]:
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a Gaussian Process regression model with an RBF kernel
kernel = 1.0 * RBF(length_scale=1.0)
gp_model = GaussianProcessRegressor(kernel=kernel, n_restarts_optimizer=10, random_state=42)

# Fit the model to the training data
gp_model.fit(X_train, y_train)

# Predict on the testing data
predictions, std_dev = gp_model.predict(X_test, return_std=True)

# Calculate mean squared error
mse = mean_squared_error(y_test, predictions)
print("Mean Squared Error:", mse)

Mean Squared Error: 13.852615652489435


## Polynomial Rregressor

In [48]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Generate polynomial features
degree = 3  # degree of the polynomial
poly = PolynomialFeatures(degree=degree)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

# Create a Polynomial Regression model
poly_model = LinearRegression()

# Fit the model to the training data
poly_model.fit(X_train_poly, y_train)

# Predict on the testing data
predictions = poly_model.predict(X_test_poly)

# Calculate mean squared error
mse = mean_squared_error(y_test, predictions)
print("Mean Squared Error:", mse)


Mean Squared Error: 35.03613626935759


## Neural Network Regression

In [49]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from sklearn.metrics import mean_squared_error

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Normalize the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Create a neural network model
model = Sequential([
    Dense(64, activation='relu', input_shape=(X_train_scaled.shape[1],)),
    Dense(32, activation='relu'),
    Dense(1)
])

# Compile the model
model.compile(optimizer=Adam(), loss='mean_squared_error')

# Fit the model to the training data
history = model.fit(X_train_scaled, y_train, epochs=100, batch_size=32, validation_split=0.2)

# Predict on the testing data
predictions = model.predict(X_test_scaled)

# Calculate mean squared error
mse = mean_squared_error(y_test, predictions)
print("Mean Squared Error:", mse)


C:\Users\HP\AppData\Roaming\Python\Python312\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 4s 62ms/step - loss: 617.5026 - val_loss: 519.7738
Epoch 2/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 13ms/step - loss: 537.3881 - val_loss: 483.0815
Epoch 3/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - loss: 508.6672 - val_loss: 442.8839
Epoch 4/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - loss: 491.9387 - val_loss: 396.8177
Epoch 5/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - loss: 446.9218 - val_loss: 344.4478
Epoch 6/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - loss: 365.2953 - val_loss: 286.1262
Epoch 7/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step - loss: 316.0934 - val_loss: 223.0947
Epoch 8/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - loss: 242.9148 - val_loss: 162.8477
Epoch 9/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - loss: 179.2135 - val_loss: 113.1413
Epoch 10/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - loss: 102.0110 - val_loss: 79.9510
Epoch 11/100
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - loss: 85.5436 - val_loss: 62.6234
Epoch 12/10

## KNN 

In [50]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a KNN regression model
k = 5  # number of neighbors
knn_model = KNeighborsRegressor(n_neighbors=k)

# Fit the model to the training data
knn_model.fit(X_train, y_train)

# Predict on the testing data
predictions = knn_model.predict(X_test)

# Calculate mean squared error
mse = mean_squared_error(y_test, predictions)
print("Mean Squared Error:", mse)


Mean Squared Error: 35.701103999999994


## SVR

In [51]:
from sklearn.model_selection import train_test_split
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a SVM regression model
svm_model = SVR(kernel='rbf')  # 'rbf' kernel is commonly used for non-linear regression

# Fit the model to the training data
svm_model.fit(X_train, y_train)

# Predict on the testing data
predictions = svm_model.predict(X_test)

# Calculate mean squared error
mse = mean_squared_error(y_test, predictions)
print("Mean Squared Error:", mse)


Mean Squared Error: 67.9756529477102
